### PR TITLE
[plugin.video.arloview] 1.0.0

### PR DESCRIPTION
**Pull request by JavaWiz1**
### Description

Stream your Arlo security camera(s) to Kodi interface.
### Checklist:

- [x ] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x ] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [x ] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practice but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-plugins/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+            xbmcplugin.addDirectoryItem(self._handle, url, list_item, is_folder)
+        # Add a sort method for the virtual folder items (alphabetically, ignore articles)
+        xbmcplugin.addSortMethod(self._handle, xbmcplugin.SORT_METHOD_LABEL_IGNORE_THE)
+        # Finish creating a virtual folder.
+        xbmcplugin.endOfDirectory(self._handle)
+
+        self.log("  Setup refresh timer", xbmc.LOGDEBUG)
+        # Update the gui to reflect signal and battery strength
+        gui_update_timer = threading.Timer(120, self._refresh_gui)
+        gui_update_timer.start()
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: This will block the python script execution from finishing for 2 minutes. That's not allowed.

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        self.log("END   main_menu()", xbmc.LOGDEBUG)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+    def _refresh_gui(self):
+        self.log("Refresh container UI", xbmc.LOGDEBUG)
+        xbmc.executebuiltin('Container.Refresh')
+
+    def _get_arlo_cameras(self):
+        self.log("_get_arlo_cameras()", xbmc.LOGDEBUG)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        if self.cameras is None:
+            self.log("Retrieving cameras...", xbmc.LOGNOTICE)
+            self.cameras = self.arlo.GetDevices("camera")
+            self._update_arlo_cameras_details()
+            if self.debug_mode:
+                self.log(json.dumps(self.cameras, indent=4), xbmc.LOGNOTICE)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+                    if camera['deviceId'] == camera_detail['serialNumber']:
+                        camera['batteryLevel'] = camera_detail['batteryLevel']
+                        camera['signalStrength'] = camera_detail['signalStrength']
+                        camera['hwVersion'] = camera_detail['hwVersion']
+                        camera['swVersion'] = camera_detail['swVersion']
+                        self.cameras[idx] = camera
+                        break
+
+    def _get_camera(self, device_id):
+        self.log("BEGIN _get_camera({})".format(device_id), xbmc.LOGDEBUG)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        tgt_camera = None
+        for camera in self._get_arlo_cameras():
+            if device_id == camera["deviceId"]:
+                tgt_camera = camera
+                break
+        self.log("END   _get_camera() returns: {}".format(tgt_camera['deviceName']), xbmc.LOGDEBUG)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+                  "Model ID  : {1}\n" + \
+                  "Batt Lvl  : {2}\n" + \
+                  "Sig Str   : {3}"
+        return fmt_str.format(camera['deviceId'],
+                              camera['properties']['modelId'],
+                              camera['batteryLevel'],
+                              camera['signalStrength'])
+
+    def _get_camera_snapshot(self, device_id):
+        self.log("BEGIN _get_camera_snapshot({})".format(device_id), xbmc.LOGDEBUG)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+                    timed_out = True
+
+        while is_playing:
+            self.log("Player is playing...", xbmc.LOGDEBUG)
+            time.sleep(1)
+            is_playing = xbmc.Player().isPlaying()
+            if not is_playing:
+                self.log("Player has stopped.", xbmc.LOGDEBUG)
+
+        self._update_arlo_cameras_details()
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: Why is this needed?

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+        self.log("END   play_camera({})".format(device_id), xbmc.LOGNOTICE)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+            self.log(json.dumps(self.basestation, indent=4), xbmc.LOGNOTICE)
+        self.log("END   arlo_login()", xbmc.LOGDEBUG)
+
+    def arlo_logout(self):
+        self.log("BEGIN arlo_logout()", xbmc.LOGDEBUG)
+        self.arlo.Logout()
+        self.log("END   arlo_logout()", xbmc.LOGDEBUG)
+
+    def run(self):
+        self.log("BEGIN run()", xbmc.LOGNOTICE)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+        self.log("  arg[0]  : {}".format(self._args[0]), xbmc.LOGNOTICE)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+        self.log("  arg[1]  : {}".format(self._args[2]), xbmc.LOGNOTICE)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+        self.log("  arg[2:n]: {}".format(self._args[2][1:]), xbmc.LOGNOTICE)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/arlo_stream.py [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+
+            self.arlo_login()
+            if _cam_id is None:
+                self.main_menu()
+                self.arlo_logout()
+            else:
+                self.play_camera(_cam_id)
+                self.arlo_logout()
+
+        self.log("END   run()", xbmc.LOGNOTICE)
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/arlo_stream.py]: logspam, please remove

File: plugin.video.arloview/resources/lib/monotonic.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
+# -*- coding: utf-8 -*-
```
Comment (Rechi) [inline plugin.video.arloview/resources/lib/monotonic.py]: This shoule be split into a script.module.

File: plugin.video.arloview/resources/lib/arlo_stream.py (hunk 4 above)
Comment (JavaWiz1) [reply plugin.video.arloview/resources/lib/arlo_stream.py]: This only occurs if user has turned on plugin debug mode in settings.  Will be used only if users have problems and will assist in troubleshooting issues.

File: plugin.video.arloview/resources/lib/arlo_stream.py (hunk 8 above)
Comment (JavaWiz1) [reply plugin.video.arloview/resources/lib/arlo_stream.py]: This updates the camera details after a streaming session (i.e. Battery level, signal strength,...)

File: plugin.video.arloview/resources/lib/monotonic.py (hunk 15 above)
Comment (JavaWiz1) [reply plugin.video.arloview/resources/lib/monotonic.py]: This is a separate module, not sure what you are asking me to do...

**Comment by JavaWiz1** [comment]
I've updated most of the requested changes.  There are a few debug statements left that will be used for troubleshooting if there are issues.
New to this GitHub and opensource development, feel free to comment if I'm not approaching this properly.

File: plugin.video.arloview/resources/lib/arlo_stream.py (hunk 13 above)
Comment (JavaWiz1) [reply plugin.video.arloview/resources/lib/arlo_stream.py]: changed to DEBUG for troubleshooting

File: plugin.video.arloview/addon.xml [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    <summary lang="en">Stream your Arlo cameras in Kodi</summary>
+    <description lang="en">List and stream each of your Arlo cameras in Kodi.  This plugin is 
+    compatible with Kodi v18 (Leia) and above.
+    
+    Thanks to Jeffrey D. Walter for creating the python interface 
+    to Arlo devices. 
+    (https://github.com/jeffreydwalter/arlo)
+    </description>
+    <source>https://github.com/JavaWiz1/repo-plugins</source>
+    <language>en</language>
```
[the remark this answers is not in the document]
Comment (JavaWiz1) [reply plugin.video.arloview/addon.xml]: removed.

File: plugin.video.arloview/addon.xml [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    <license>Apache 2.0</license>
+    <email>JavaWiz1@hotmail.com</email>
+    <news>
+    v1.0.0 (1/1/2019)
+    - Initial release
+    </news>
+    <assets>
+        <icon>resources/icon.png</icon>
+        <fanart>resources/fanart.png</fanart>
+        <clearlogo></clearlogo>
```
[the remark this answers is not in the document]
Comment (JavaWiz1) [reply plugin.video.arloview/addon.xml]: removed

File: plugin.video.arloview/default.py [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+
+@author: adamico
+'''
+import sys
+import xbmc
+
+from resources.lib import arlo_stream
+
+if __name__ == '__main__':
+    xbmc.log("===> KodiPlugin-ArloStream START <===", xbmc.LOGNOTICE)
```
[the remark this answers is not in the document]
Comment (JavaWiz1) [reply plugin.video.arloview/default.py]: removed.

File: plugin.video.arloview/default.py [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+    arlo_stream.ArloStream(sys.argv).run()
+    xbmc.log("===> KodiPlugin-ArloStream END   <===", xbmc.LOGNOTICE)
```
[the remark this answers is not in the document]
Comment (JavaWiz1) [reply plugin.video.arloview/default.py]: removed.

File: plugin.video.arloview/resources/lib/Arlo.py [hunk 20]
```diff
@@ -0,0 +1,1085 @@
+# #
```
[the remark this answers is not in the document]
Comment (Rechi) [reply plugin.video.arloview/resources/lib/Arlo.py]: A script.module add-on is a python module packaged seperately. Then you just import it (like script.module.requests). The advantage is that it can be seperately updated and if multiple add-ons use it, they don't have to ship it themself.

You can look at the layout of one at https://github.com/xbmc/repo-scripts/tree/leia/script.module.requests. If you still need some help, I'll assist.

File: plugin.video.arloview/resources/lib/arlo_stream.py (hunk 4 above)
Comment (Rechi) [reply plugin.video.arloview/resources/lib/arlo_stream.py]: okay, then it's fine

File: plugin.video.arloview/resources/lib/arlo_stream.py (hunk 12 above)
Comment (Rechi) [reply plugin.video.arloview/resources/lib/arlo_stream.py]: If you enable debugging in Kodi, the same info is already loged from Kodi itself (`StartScript - calling plugin ...`).

File: plugin.video.arloview/resources/lib/arlo_stream.py (hunk 8 above)
Comment (Rechi) [reply plugin.video.arloview/resources/lib/arlo_stream.py]: Where are the updated details actually used?

File: plugin.video.arloview/resources/lib/Arlo.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (JavaWiz1) [reply plugin.video.arloview/resources/lib/Arlo.py]: the Arlo.py is a module written by a different author.  He is aware that I am building a Kodi plugin with his module.  How do I create addon.xml and the other addon information (website, source email)?  
As this is a specific module related to this plugin, I put it in the lib directory.

File: plugin.video.arloview/resources/lib/arlo_stream.py (hunk 12 above)
Comment (JavaWiz1) [reply plugin.video.arloview/resources/lib/arlo_stream.py]: Removed for the next pull request

File: plugin.video.arloview/resources/lib/Arlo.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (Rechi) [reply plugin.video.arloview/resources/lib/Arlo.py]: I'm aware that it is written by a different author. As long as you comply with the licence of that module all is fine.
You can look at https://github.com/xbmc/repo-scripts/compare/leia...Rechi:script.module.arlo to see that module packaged. You can just pick that for testing and submitting if it works.
